### PR TITLE
Close CPPL-75 and CPPL-76

### PR DESCRIPTION
**Pull request by sorbeee**

### CODE REVIEW

File: src/Server/libdal/Facade/idbfacade.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class IDbFacade
+{
+public:
+	using Ptr = std::shared_ptr<IDbFacade>;
+
+	virtual void AddUser(const User& user) = 0;
+	virtual std::optional<User> GetUserById(IdType id) = 0;
+	virtual std::optional<User> GetUserByLogin(const std::string& login) = 0;
+	virtual void UpdateUser(const User& user) = 0;
+	virtual void RemoveUser(IdType id) = 0;
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Facade/idbfacade.h:15]: Something wrong with code formatting.

File: src/Server/libdal/Facade/idbfacade.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    virtual void AddIncome(const Income& income) = 0;
+    virtual std::optional<Income> GetIncomeById(const IdType& id) = 0;
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Facade/idbfacade.h:18]: Something wrong with code formatting.

File: src/Server/libdal/Facade/dbfacade.h [hunk 3]
```diff
@@ -17,7 +17,19 @@ class DbFacade : public IDbFacade
 	void UpdateUser(const User &user) override;
 	void RemoveUser(IdType id) override;
 
+    void AddIncome(const Income& income) override;
+    std::optional<Income> GetIncomeById(const IdType& id) override;
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Facade/dbfacade.h:21]: Something wrong with code formatting.

File: src/Server/libdal/Facade/dbfacade.h [hunk 4]
```diff
@@ -17,7 +17,19 @@ class DbFacade : public IDbFacade
...
+    void AddList(const List& list_) override;
+    std::optional<List> GetListById(const IdType& id) override;
+    void UpdateList(const List& list_) override;
+    void RemuveList(const IdType& id) override;
+
 private:
 	pqxx::connection m_connection;
 	UserRepository m_users;
+    IncomeRepository m_incomes;
+    ListRepository m_lists;
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Facade/dbfacade.h:34]: Something wrong with code formatting.

File: src/Server/libdal/Repositories/ListRepository/ListRepository.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+       w.commit();
+    }
+    catch(const pqxx::pqxx_exception& e)
+    {
+        throw DatabaseFailure();
+    }
+}
+
+
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Repositories/ListRepository/ListRepository.cpp:90]: Something wrong with code formatting.

File: src/Server/libdal/Facade/dbfacade.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+void DbFacade::RemoveUser(IdType id)
+{
+	m_users.Remove(id);
+}
+
+void DbFacade::AddIncome(const Income &income)
+{
+    m_incomes.Add(income);
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Facade/dbfacade.cpp:35]: Something wrong with code formatting. There is different length before Remove method and Add method.

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.h [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+class IncomeRepository
+{
+public:
+    IncomeRepository(pqxx::connection& databaseConnection);
+
+    void Add(const Income& income);
+    std::optional<Income> GetIncome(const idType& id);
+    void Update(const Income& income);
+    void Remuve(const idType& id);
```
Comment (VasylDobrovolskyi) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.h:16]: Maybe you meant "remove".

File: src/Server/libdal/Facade/dbfacade.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+{
+    m_incomes.Update(income);
+}
+
+void DbFacade::RemuveIncome(const IdType &id)
+{
+    m_incomes.Remuve(id);
+}
+
+void DbFacade::AddList(const List &list_)
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Facade/dbfacade.cpp:53]: Why are you placing "_" at the end of variable name, please remove it wherever possible.

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+               return std::nullopt;
+
+        else
+        {
+            return Income{income[ID_FIELD].as<idType>(), income[USER_ID].as<idType>(),
+                          income[INCOME_NAME].as<std::string>(), income[AMOUNT].as<double>(),
+                          income[CATEGORY_ID].as<idType>(), income[ADD_TIME].as<std::string>()};
+        }
+    }
+    catch(const pqxx::pqxx_exception& e)
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:60]: Consider passing error message from pqxx_exception to the DatabaseFailure.

File: src/Server/libdal/Repositories/ListRepository/ListRepository.h [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <pqxx/pqxx>
+#include <chrono>
+
+#include "Exceptions/databasefailure.h"
+
+class ListRepository
+{
+public:
+    ListRepository(pqxx::connection& database_connection);
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/ListRepository/ListRepository.h:11]: Word database in the name of variable database_connection may be redundant. Remove it, or shorten, and also synchronize with other repositories.

**Review by Denys-Bykov-0 — CHANGES_REQUESTED** [review]

File: src/Server/libdal/Facade/dbfacade.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+{
+	m_users.Update(user);
+}
+
+void DbFacade::RemoveUser(IdType id)
+{
+	m_users.Remove(id);
+}
+
+void DbFacade::AddIncome(const Income &income)
```
Comment (AndriyFedorenko) [inline src/Server/libdal/Facade/dbfacade.cpp:33]: you write `type &name;` some time `type& mane;` please use the same style everywhere

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+IncomeRepository::IncomeRepository(pqxx::connection& databaseConnection) : m_databaseConnection(databaseConnection)
+{
+}
+
+void IncomeRepository::Add(const Income& income)
+{
+    pqxx::work w{m_databaseConnection};
+
+    w.exec0("INSERT INTO " + TABLE_NAME + " (" + USER_ID + ", " + INCOME_NAME + ", " + AMOUNT + ", "
```
Comment (AndriyFedorenko) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:24]: Why don’t you handle exceptions?

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            + w.quote(income.categoryId) + ", " + w.quote(income.addTime) + ", "
+            + w.quote(income.expoirationTime) + ");"
+             );
+
+     w.commit();
+}
+
+std::optional<Income> IncomeRepository::GetIncome(const idType& id)
+{
+    pqxx::work w{m_databaseConnection};
```
Comment (AndriyFedorenko) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:36]: Work can also throw exception if connection is invalid for instance.

File: src/Server/libdal/Repositories/ListRepository/ListRepository.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    const std::string& TABLE_NAME = "List";
+    const std::string& ID_FIELD = "id";
+    const std::string& USER_ID = "userId";
+    const std::string& LIST_NAME = "name";
+    const std::string& LIST_STATE_ID = "stateId";
+    const std::string& TABLE_WITH_CATEGORY = "ListState";
+}
+
+
+ListRepository::ListRepository(pqxx::connection& database_connection) : m_database_connection(database_connection)
```
Comment (AndriyFedorenko) [inline src/Server/libdal/Repositories/ListRepository/ListRepository.cpp:14]: Please add const

File: src/Server/libdal/DTOs/Income.h [hunk 15]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <string>
+#include "Types.h"
+
+struct Income
+{
+	IdType income_id;
+	IdType user_id;
+    std::string name;
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/DTOs/Income.h:10]: Indentation issue.

File: src/Server/libdal/Facade/dbfacade.cpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
+#include "dbfacade.h"
+
+DbFacade::DbFacade(const std::string& connection_string) : m_connection(connection_string), m_users(m_connection),
+														   m_incomes(m_connection), m_lists(m_connection),
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Facade/dbfacade.cpp:4]: Indentation issue. Please reformat initializer list to look like this:

ListSubPage::ListSubPage(const QString& name, int id, QWidget *parent)
	: QWidget(parent)
	, m_ui(new Ui::ListSubPage)
	, m_name(name)
	, m_id(id)
{

}

File: src/Server/libdal/Facade/dbfacade.cpp [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+{
+	return m_users.GetById(id);
+}
+
+std::optional<User> DbFacade::GetUserByLogin(const std::string& login)
+{
+	return m_users.GetByLogin(login);
+}
+
+void DbFacade::UpdateUser(const User &user)
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Facade/dbfacade.cpp:26]: Write Type& var_name everywhere.

File: src/Server/libdal/DTOs/IncomeCategory.h [hunk 18]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <string>
+#include "Types.h"
+
+struct IncomeCategory
+{
+	IdType income_category_id;
+    std::string name;
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/DTOs/IncomeCategory.h:9]: Indentation issue.

File: src/Server/libdal/Repositories/IncomeCategoryRepository/IncomeCategoryRepository.cpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            income_categories.push_back(ParseSQLRow(row));
+        }
+        return income_categories;
+    }
+    catch(const pqxx::pqxx_exception& e)
+    {
+		throw DatabaseFailure();
+    }
+}
+
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/IncomeCategoryRepository/IncomeCategoryRepository.cpp:61]: So many blank lines at the end of the file, for what?

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+    }
+}
+
+std::optional<Income> IncomeRepository::GetIncome(const IdType& income_id)
+{
+	try
+    {
+        pqxx::work w{m_db_connection};
+		pqxx::row income = w.exec1("SELECT * FROM " + TABLE_NAME + " WHERE " + ID_FIELD + " = "  + w.quote(income_id) + ";");
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:49]: Indentation.

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+        if(!income[EXPERATION_TIME].is_null())
+        {
+			return Income{income[ID_FIELD].as<IdType>(), income[USER_ID].as<IdType>(),
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:53]: Really bad indentation issue.

File: src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+		income.name = row[INCOME_NAME].as<std::string>();
+        income.amount = row[AMOUNT].as<double>();
+		income.category_id = row[CATEGORY_ID].as<IdType>();
+		income.add_time = row[ADD_TIME].as<std::string>();
+		income.expoiration_time = "";
+
+        return income;
+    }
+}
+
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/IncomeRepository/IncomeRepository.cpp:161]: Too many blank lines.

File: src/Server/libdal/Repositories/ListRepository/ListRepository.cpp [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+std::optional<List> ListRepository::GetList(const IdType& list_id)
+{
+    try
+    {
+        pqxx::work w{m_db_connection};
+        pqxx::row list =  w.exec1("SELECT * FROM " + TABLE_NAME + " WHERE " + ID_FIELD + " = " + w.quote(list_id));
+
+        if(!list.empty())
+        {
+			return List{list[ID_FIELD].as<IdType>(), list[USER_ID].as<IdType>(),
```
Comment (Denys-Bykov-0) [inline src/Server/libdal/Repositories/ListRepository/ListRepository.cpp:60]: Another indentation issue.

**Comment by yaroslav-kernitskyi** [comment]
dbfacade.cpp was not deleted

File: src/Server/libdal/Facade/dbfacade.cpp (hunk 16 above)
Comment (yaroslav-kernitskyi) [reply src/Server/libdal/Facade/dbfacade.cpp:4]: @Denys-Bykov-0, consider using ctrl+e or <> button on tool panel to format code. It looks better and it is easier to copy.
```
ListSubPage::ListSubPage(const QString& name, int id, QWidget *parent)
	: QWidget(parent)
	, m_ui(new Ui::ListSubPage)
	, m_name(name)
	, m_id(id)
{

}
```

**Review by Denys-Bykov-0 — APPROVED** [review]

**Review by VasylDobrovolskyi — APPROVED** [review]

File: src/Server/libdal/Exceptions/DatabaseFailure.h [hunk 24]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <stdexcept>
+#include "pqxx/pqxx"
```
Comment (yaroslav-kernitskyi) [inline src/Server/libdal/Exceptions/DatabaseFailure.h:4]: Do you really need it here?

**Review by yaroslav-kernitskyi — APPROVED** [review]

**Pull request merged**